lib/pos_handshake_controller.go

```diff
@@ -0,0 +1,160 @@
+func (hc *HandshakeController) InitiateHandshake(rn *RemoteNode) {
+	nonce := uint64(RandInt64(math.MaxInt64))
+	if err := rn.InitiateHandshake(nonce); err != nil {
+		glog.Errorf("RemoteNode.InitiateHandshake: Error initiating handshake: %v", err)
+		hc.rnManager.Disconnect(rn)
+	}
+	hc.usedNonces.Add(nonce)
+}
+
+// _handleHandshakeCompleteMessage handles HandshakeComplete control messages, sent by RemoteNodes.
+func (hc *HandshakeController) _handleHandshakeCompleteMessage(origin *Peer, desoMsg DeSoMessage) {
+	if desoMsg.GetMsgType() != MsgTypePeerHandshakeComplete {
+		return
+	}
+
+	// Get the handshake information of this peer.
+	remoteNode := hc.rnManager.GetRemoteNodeFromPeer(origin)
+	if remoteNode == nil {
+		return
+	}
+
+	if remoteNode.GetNegotiatedProtocolVersion() == ProtocolVersion2 {
+		if err := hc.handleHandshakeCompletePoSMessage(remoteNode); err != nil {
+			glog.Errorf("HandshakeController._handleHandshakeCompleteMessage: Error handling PoS handshake peer message: %v", err)
+			hc.rnManager.Disconnect(remoteNode)
+			return
+		}
+	}
+	hc.rnManager.ProcessCompletedHandshake(remoteNode)
+}
+
+func (hc *HandshakeController) handleHandshakeCompletePoSMessage(remoteNode *RemoteNode) error {
+
+	validatorPk := remoteNode.GetValidatorPublicKey()
+	// If the remote node is not a potential validator, we don't need to do anything.
+	if validatorPk == nil {
+		return nil
+	}
+
+	// Lookup the validator in the ValidatorIndex with the same public key.
+	existingValidator, ok := hc.rnManager.GetValidatorIndex().Get(validatorPk.Serialize())
+	// For inbound RemoteNodes, we should ensure that there isn't an existing validator connected with the same public key.
+	// Inbound nodes are not initiated by us, so we shouldn't have added the RemoteNode to the ValidatorIndex yet.
+	if remoteNode.IsInbound() && ok {
+		return fmt.Errorf("HandshakeController.handleHandshakeCompletePoSMessage: Inbound RemoteNode with duplicate validator public key")
+	}
+	// For outbound RemoteNodes, we have two possible scenarios. Either the RemoteNode has been initiated as a validator,
+	// in which case it should already be in the ValidatorIndex. Or the RemoteNode has been initiated as a regular node,
+	// in which case it should not be in the ValidatorIndex, but in the NonValidatorOutboundIndex. So to ensure there is
+	// no duplicate connection with the same public key, we only check whether there is a validator is in the ValidatorIndex
+	// with the RemoteNode's public key. If there is one, we want to ensure that these two RemoteNodes have identical ids.
+	if remoteNode.IsOutbound() && ok {
+		if remoteNode.GetId() != existingValidator.GetId() {
+			return fmt.Errorf("HandshakeController.handleHandshakeCompletePoSMessage: Outbound RemoteNode with duplicate validator public key. "+
+				"Existing validator id: %v, new validator id: %v", existingValidator.GetId().ToUint64(), remoteNode.GetId().ToUint64())
+		}
+	}
+	return nil
```

Comment:
nesting
```suggestion
	if !remoteNode.IsOutbound() || !ok {
		return
	}
	
	if remoteNode.GetId() != existingValidator.GetId() {
		return fmt.Errorf("HandshakeController.handleHandshakeCompletePoSMessage: Outbound RemoteNode with duplicate validator public key. "+
			"Existing validator id: %v, new validator id: %v", existingValidator.GetId().ToUint64(), remoteNode.GetId().ToUint64())
	}
```

Reply:
The first one was cool. But this one looks a little unintuitive to me. The nested condition nicely parallels the previous `IsInbound` check. The `!remoteNode.IsOutbound() || !ok` it less clear to me.